command/sync.go

```diff
@@ -302,46 +302,105 @@ func (s Sync) getSourceAndDestinationObjects(ctx context.Context, srcurl, dsturl
 	}
 
 	var (
-		sourceObjects []*storage.Object
-		destObjects   []*storage.Object
-		wg            sync.WaitGroup
+		sourceObjects = make(chan *storage.Object, extsortChannelBufferSize)
+		destObjects   = make(chan *storage.Object, extsortChannelBufferSize)
 	)
 
+	extsortDefaultConfig := extsort.DefaultConfig()
+	extsortConfig := &extsort.Config{
+		ChunkSize:          extsortChunkSize,
+		NumWorkers:         extsortDefaultConfig.NumWorkers,
+		ChanBuffSize:       extsortChannelBufferSize,
+		SortedChanBuffSize: extsortChannelBufferSize,
+	}
+	extsortDefaultConfig = nil
+
 	// get source objects.
-	wg.Add(1)
 	go func() {
-		defer wg.Done()
-		srcObjectChannel := sourceClient.List(ctx, srcurl, s.followSymlinks)
-		for srcObject := range srcObjectChannel {
-			if s.shouldSkipObject(srcObject, true) {
-				continue
+		defer close(sourceObjects)
+		unfilteredSrcObjectChannel := sourceClient.List(ctx, srcurl, s.followSymlinks)
+		filteredSrcObjectChannel := make(chan extsort.SortType, extsortChannelBufferSize)
+
+		go func() {
+			defer close(filteredSrcObjectChannel)
+			// filter and redirect objects
+			for st := range unfilteredSrcObjectChannel {
+				if s.shouldSkipObject(st, true) {
+					continue
+				}
+				filteredSrcObjectChannel <- *st
 			}
-			sourceObjects = append(sourceObjects, srcObject)
+		}()
+
+		var (
+			sorter        *extsort.SortTypeSorter
+			srcOutputChan chan extsort.SortType
+		)
+
+		sorter, srcOutputChan, srcErrCh := extsort.New(filteredSrcObjectChannel, storage.FromBytes, storage.Less, extsortConfig)
+		sorter.Sort(ctx)
+
+		for srcObject := range srcOutputChan {
+			o := srcObject.(storage.Object)
+			sourceObjects <- &o
 		}
+
+		// read and print the external sort errors
+		go func() {
+			for err := range srcErrCh {
+				printError(s.fullCommand, s.op, err)
+			}
+		}()
 	}()
 
 	// get destination objects.
-	wg.Add(1)
 	go func() {
-		defer wg.Done()
-		destObjectsChannel := destClient.List(ctx, destObjectsURL, false)
-		for destObject := range destObjectsChannel {
-			if s.shouldSkipObject(destObject, false) {
-				continue
+		defer close(destObjects)
+		unfilteredDestObjectsChannel := destClient.List(ctx, destObjectsURL, false)
+		filteredDstObjectChannel := make(chan extsort.SortType, extsortChannelBufferSize)
+
+		go func() {
+			defer close(filteredDstObjectChannel)
+
+			// filter and redirect objects
+			for dt := range unfilteredDestObjectsChannel {
+				if s.shouldSkipObject(dt, true) {
```

Comment:
Previous implementation sets `verbose` flag to false. 

If verbose flag is set to true and  destination prefix is empty `s5cmd sync` prints `no object found` error message.
Could you please write a test that also checks stderr for sync command?

Reply:
Ok, thanks for warning.

ps. It is done with  https://github.com/peak/s5cmd/pull/483/commits/171d607bb8f29c7eb18384334f99303dc9de9acb